src/components/CharacterAssignment.jsx

```diff
@@ -165,27 +164,21 @@ const CharacterAssignment = ({ actors, currentScriptId }) => {
       return;
     }
     // Get the column ids of the source and destination columns
-    // todo Fix columns not being fully cloned before modifying and calling setColumns() with the new data.
-    // todo Columns is a nested object and { ...columns } only clones the first level properties.
-    // todo Also, start and finish below directly reference the values in columns.
-    // todo Changing them changes columns directly (and violates the "don't change state directly" rule of react).
-    const start = columns[source.droppableId];
-    const finish = columns[destination.droppableId];
+    // We create a structured clone to make sure we aren't changing the state directly.
+    const start = structuredClone(columns[source.droppableId]);
+    const finish = structuredClone(columns[destination.droppableId]);
 
     // Remove the character from the column it was picked up from and update the column values
     const index = start.characterIds.indexOf(draggableId);
     start.characterIds.splice(index, 1);
-
     // Add the character to the column it was dropped on
     finish.characterIds.push(draggableId);
 
     // Update the assigned character in the database and if it works, reset the column values to represent the assignment
     const response = await assignCharacter(draggableId, destination.droppableId, currentScriptId);
-    if (response.success) {
-      // todo set columns BEFORE awaiting database change to avoid visual jitter on page due to delay
-      // todo revert columns to original state if assignCharacter() fails
-      setColumns({ ...columns, [start.id]: start, [finish.id]: finish });
-    } else {
+    setColumns({ ...columns, [start.id]: start, [finish.id]: finish });
```

Comment:
This needs to happen before the await for the flickering to disappear

Reply:
Right, fixed it! Tested that it resets to the previous value if the database update changes too. Thanks for remembering this!